Adds live migration procedures for sdn to ovnk (#76672)
openshift/openshift-docs: modules/nw-ovn-kubernetes-live-migration-about.adoc

// Module included in the following assemblies:
//
// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc

[id="nw-ovn-kubernetes-live-migration-about_{context}"]
= Live migration to the OVN-Kubernetes network plugin overview

> The live migration method is the process in which the OpenShift SDN network plugin and its network configurations, connections, and associated resources, are migrated to the OVN-Kubernetes network plugin without service interruption. It is available for {product-title}, {product-dedicated}, {product-rosa}, and Azure Red Hat OpenShift deployment types It is not available for HyperShift deployment types. This migration method is valuable for deployment types that require constant service availability and offers the following benefits:

Reviewer: nit: `deployment types. It is not available ...`

* Continuous service availability
* Minimized downtime
* Automatic node rebooting
* Seamless transition from the OpenShift SDN network plugin to the OVN-Kubernetes network plugin.

Although a rollback procedure is provided, the live migration is intended to be a one-way process.

////
[IMPORTANT]
====

> Migrating to or from the OVN-Kubernetes network plugin is not supported for managed OpenShift cloud services such as {product-dedicated}, Azure Red Hat OpenShift(ARO), and Red Hat OpenShift Service on AWS (ROSA).

Reviewer: We can remove this line.

====
////

include::snippets/sdn-deprecation-statement.adoc[]

The following sections provide more information about the live migration method.

[id="supported-platforms-live-migrating-ovn-kubernetes"]
== Supported platforms when using the live migration method

> The following table provides information about the supported platforms for the live migration type.

Reviewer: We shall mention that both IPI and UPI deployments are supported.

Author: @pliurh Is it 

```suggestion
|===
| Platform              | Live Migration

| Bare metal hardware (IPI and UPI)             |&#10003;
```

OR 
```
|===
| Platform              | Live Migration
...
| IPI             |&#10003;
| UPI            |&#10003;
```

Reviewer: We support both IPI and UPI for all platforms. So the first one looks better, but not only for baremetal.

.Supported platforms for the live migration method
[cols="1,1", options="header"]
|===
| Platform              | Live Migration

| Bare metal hardware             |&#10003;
| Amazon Web Services (AWS)       |&#10003;
| Google Cloud Platform (GCP)     |&#10003;
| {ibm-cloud-name}                |&#10003;
| Microsoft Azure                 |&#10003;
| {rh-openstack-first}            |&#10003;
| VMware vSphere                  |&#10003;
| AliCloud                        |&#10003;
| Nutanix                         |&#10003;
|===

[id="considerations-live-migrating-ovn-kubernetes-network-provider_{context}"]
== Considerations for live migration to the OVN-Kubernetes network plugin

Cluster administrators should consider the following before using the live migration method to the OVN-Kubernetes network plugin:

> * The live migration procedure is unsupported for clusters with OpenShift SDN multitenant mode enabled. You must disable multitenant mode before beginning the process.

Reviewer 2: For multitenant mode cluster, I don't think it can be disalbed. so we just say it's unsupported.

```suggestion
* The live migration procedure is unsupported for clusters with OpenShift SDN multitenant mode enabled. 
```

> * When the live migration procedure begins, multitenant isolation is permanently disabled. Multitenant isolation is only supported on OpenShift SDN.

Reviewer 2: this line can be removed

```suggestion

```

* Egress routers block the live migration process. This feature must be removed before beginning the live migration process.

* During the live migration, multicast, egress IP addresses, and egress firewalls are temporarily disabled. They can be migrated from OpenShift SDN to OVN-Kubernetes after the live migration process has finished.

* Although a rollback procedure is provided, the migration is intended to be a one-way process. For users that want to rollback to OpenShift-SDN, migration from OpenShift-SDN to OVN-Kubernetes must have succeeded. Users can follow the same procedure below to migrate to the OpenShift SDN network plugin from the OVN-Kubernetes network plugin.

* The live migration not supported on HyperShift clusters.

> * New nodes should not be created during the live migration because both CNIs would try to setup the new node.

Reviewer: We can remove this line. Adding nodes is supported in 4.16.

* OpenShift SDN does not support IPsec. After migration, cluster administrators can enable IPsec.

> * OpenShift SDN does not support IPv6. Consequently, live migration to OVN-Kubernetes with IPv6 is not supported. Cluster administrators can enable IPv4, IPv6, or dual-stack if necessary after migration is completed.

Reviewer: Cluster administrators can enable IPv6 single-stack, or dual-stack if necessary after migration is completed.

Author: Changed to 

```suggestion
* OpenShift SDN does not support IPv6. Consequently, live migration to OVN-Kubernetes with IPv6 is not supported. Cluster administrators can enable IPv4, IPv6 single-stack, or dual-stack if necessary after migration is completed.
```

Do you want the mentioning of IPv4 removed entirely?

Reviewer: I was wrong. You cannot migrate from IPv4 single stack to IPv6 single stack. So here we shall only mention dual-stack. We don't neet to mention IPv4 single stack, once the migration is completed, the cluster is already IPv4 single stack.

Author: @pliurh how is this? 

* OpenShift SDN does not support IPv6. Consequently, live migration to OVN-Kubernetes with IPv6 is not supported. Cluster administrators can enable dual-stack if necessary after migration is completed.

* The cluster MTU is the MTU value for pod interfaces. It is always less than your hardware MTU to account for the cluster network overlay overhead. The overhead is 100 bytes for OVN-Kubernetes and 50 bytes for OpenShift SDN.
+
During the live migration, both OVN-Kubernetes and OpenShift SDN run in parallel. OVN-Kubernetes manages the cluster network of some nodes, while OpenShift SDN manages the cluster network of others. To ensure that cross-CNI traffic remains functional, the Cluster Network Operator updates the routable MTU to ensure that both CNIs share the same overlay MTU. As a result, after the migration has completed, the cluster MTU is 50 bytes less.

* Some parameters of OVN-Kubernetes cannot be changed after installation. The following parameters can be set before starting the live migration:

** `v4InternalSubnet`

> ** `v4TransitSwitchSubnet`

Reviewer: @martinkennelly Do you know whether we will backport this API change to 4.15?

> * Unless otherwise configured, OVN-Kubernetes uses subnets `100.64.0.0/16`, `100.88.0.0/16`, and `169.254.169.0/29` internally by default. Cluster administrators must ensure that these addresses do not overlap with any existing subnets in their network. If there is an overlap, administrators must explicitly set the `v4InternalSubnet` and `v4TransitSwitchSubnet` before initiating the live migration.

Reviewer 2: I guess we need to tell, 
`100.64.0.0/1` is for internalJoinSubnet by default
`100.88.0.0/16` is for internalTransitSwitchSubnet by default

if the above CIDR is used for other network, please update the another CIDR for ovn. eg.  `100.64.0.0/16` is used.  then we can change internalJoinSubnet to `100.63.0.0/16` before migration.  command
```
oc patch network.operator.openshift.io cluster --type='merge' -p='{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"ipv4":{"internalJoinSubnet": "100.63.0.0/16"}}}}}'
```

if the default internalTransitSwitchSubnet 100.88.0.0/16 is used for other network, we can change another CIDR for OVN. eg 100.99.0.0/16. command:
```
oc patch network.operator.openshift.io cluster --type='merge' -p='{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"ipv4":{"internalTransitSwitchSubnet": "100.99.0.0/16"}}}}}'
```

> * In most cases, the live migration is independent to the secondary interfaces of pods created by the Multus CNI plugin. However, if these secondary interfaces were set up on the default network interface controller (NIC) of the host, for example, using macvlan or bridge interfaces with the default NIC as the control node, OVN-Kubernetes might encounter malfunctions. Users should remove such configurations before proceeding with the live migration.

Reviewer: One more caveat shall be added. When there are multiple NICs in the host, and the default route is not on the interface that has the Kubernetes NodeIP. Users shall use the offline migration instead.

// Module included in the following assemblies:
//
// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc

[id="nw-ovn-kubernetes-live-migration-about_{context}"]
= Live migration to the OVN-Kubernetes network plugin overview

> The live migration method is the process in which the OpenShift SDN network plugin and its network configurations, connections, and associated resources, are migrated to the OVN-Kubernetes network plugin without service interruption. It is available for {product-title}, {product-dedicated}, {product-rosa}, and Azure Red Hat OpenShift deployment types. It is not available for HyperShift deployment types. This migration method is valuable for deployment types that require constant service availability and offers the following benefits:

Reviewer 3: ```suggestion
The live migration method is the process in which the OpenShift SDN network plugin and its network configurations, connections, and associated resources, are migrated to the OVN-Kubernetes network plugin without service interruption. It is available for {product-title}, {product-dedicated}, {product-rosa}, and Azure Red Hat OpenShift deployment types. It is not available for HyperShift deployment types. For more information on other limitations and cosiderations, see the " Considerations for live migration to the OVN-Kubernetes network plugin" section that follows.

This migration method is valuable for deployment types that require constant service availability and offers the following benefits:
```

* Continuous service availability
* Minimized downtime
* Automatic node rebooting
* Seamless transition from the OpenShift SDN network plugin to the OVN-Kubernetes network plugin

> Although a rollback procedure is provided, the live migration is intended to be a one-way process.

Reviewer 3: ```suggestion
The migration is intended to be a one-way process. However, a rollback procedure is provided, if needed.
```

include::snippets/sdn-deprecation-statement.adoc[]

The following sections provide more information about the live migration method.

[id="supported-platforms-live-migrating-ovn-kubernetes"]
== Supported platforms when using the live migration method

The following table provides information about the supported platforms for the live migration type.

.Supported platforms for the live migration method
[cols="1,1", options="header"]
|===
| Platform              | Live Migration

| Bare metal hardware (IPI and UPI)             |&#10003;
| Amazon Web Services (AWS) (IPI and UPI)       |&#10003;
| Google Cloud Platform (GCP) (IPI and UPI)     |&#10003;
| {ibm-cloud-name} (IPI and UPI)                |&#10003;
| Microsoft Azure (IPI and UPI)                 |&#10003;
| {rh-openstack-first} (IPI and UPI)            |&#10003;
| VMware vSphere (IPI and UPI)                  |&#10003;
| AliCloud (IPI and UPI)                        |&#10003;
| Nutanix (IPI and UPI)                         |&#10003;
|===

[id="considerations-live-migrating-ovn-kubernetes-network-provider_{context}"]
== Considerations for live migration to the OVN-Kubernetes network plugin

> Cluster administrators should consider the following before using the live migration method to the OVN-Kubernetes network plugin:

Reviewer 3: ```suggestion
Before using the live migration method to the OVN-Kubernetes network plugin, cluster administrators should consider the following conditions:
```

* The live migration procedure is unsupported for clusters with OpenShift SDN multitenant mode enabled. 

* Egress router pods block the live migration process. They must be removed before beginning the live migration process.

* During the live migration, multicast, egress IP addresses, and egress firewalls are temporarily disabled. They can be migrated from OpenShift SDN to OVN-Kubernetes after the live migration process has finished.

> * Although a rollback procedure is provided, the migration is intended to be a one-way process. For users that want to rollback to OpenShift-SDN, migration from OpenShift-SDN to OVN-Kubernetes must have succeeded. Users can follow the same procedure below to migrate to the OpenShift SDN network plugin from the OVN-Kubernetes network plugin.

Reviewer 3: ```suggestion
* The migration is intended to be a one-way process. However, for users that want to rollback to OpenShift-SDN, migration from OpenShift-SDN to OVN-Kubernetes must have succeeded. Users can follow the same procedure below to migrate to the OpenShift SDN network plugin from the OVN-Kubernetes network plugin.
```

> * The live migration not supported on HyperShift clusters.

Reviewer 3: ```suggestion
* The live migration is not supported on HyperShift clusters.
```

> * OpenShift SDN does not support IPsec. After migration, cluster administrators can enable IPsec.

Reviewer 3: ```suggestion
* OpenShift SDN does not support IPsec. After the migration, cluster administrators can enable IPsec.
```

> * OpenShift SDN does not support IPv6. Cluster administrators can enable dual-stack if necessary after migration is completed.

Reviewer 3: I like parallelism in lists. Personal preference.
```suggestion
* OpenShift SDN does not support IPv6. After the migration, cluster administrators can enable dual-stack if necessary after migration is completed.
```

> * OpenShift SDN does not support IPv6. Consequently, live migration to OVN-Kubernetes with IPv6 is not supported. Cluster administrators can enable dual-stack if necessary after migration is completed.

Reviewer 4: Seems we do need this statement - `Consequently, live migration to OVN-Kubernetes with IPv6 is not supported`

* The cluster MTU is the MTU value for pod interfaces. It is always less than your hardware MTU to account for the cluster network overlay overhead. The overhead is 100 bytes for OVN-Kubernetes and 50 bytes for OpenShift SDN.
+
During the live migration, both OVN-Kubernetes and OpenShift SDN run in parallel. OVN-Kubernetes manages the cluster network of some nodes, while OpenShift SDN manages the cluster network of others. To ensure that cross-CNI traffic remains functional, the Cluster Network Operator updates the routable MTU to ensure that both CNIs share the same overlay MTU. As a result, after the migration has completed, the cluster MTU is 50 bytes less.

> * Some parameters of OVN-Kubernetes cannot be changed after installation. The following parameters can be set before starting the live migration:

Reviewer 3: Is this change correct?
```suggestion
* Some parameters of OVN-Kubernetes cannot be changed after installation. The following parameters can only be set before starting the live migration:
```

** `InternalTransitSwitchSubnet`
** `internalJoinSubnet`

* Unless otherwise configured, OVN-Kubernetes uses the following IP address ranges:

> ** `100.64.0.0/1`. This IP address range is used for `internalJoinSubnet` of OVN-Kubernetes by default. If this IP address range is already in use, enter the following command to update it to `100.63.0.0/16`:

Reviewer 3: ```suggestion
** `100.64.0.0/1`. This IP address range is used for the `internalJoinSubnet` of OVN-Kubernetes by default. If this IP address range is already in use, enter the following command to update it to `100.63.0.0/16`:
```

> ** `100.64.0.0/1`. This IP address range is used for `internalJoinSubnet` of OVN-Kubernetes by default. If this IP address range is already in use, enter the following command to update it to `100.63.0.0/16`:

Reviewer 4: Suggest update it to be` If this IP address range is already in use in OpenShift SDN`

Reviewer: OVN-K shall not use this subnet if it's already used by any of the customer's networks, not just by OpenShiftSDN.

Author: @weliang1 @pliurh How about this? 

```suggestion
** `100.64.0.0/1`. This IP address range is used for `internalJoinSubnet` of OVN-Kubernetes by default. If this IP address range is already in use by another network, enter the following command to update it to 100.63.0.0/16:
```
Not sure if it needs changed at all based on Peng's comment.

+
[source,terminal]
----
$ oc patch network.operator.openshift.io cluster --type='merge' -p='{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"ipv4":{"internalJoinSubnet": "100.63.0.0/16"}}}}}'
----

> ** `100.88.0.0/16`. This IP address range is used for `internalTransSwitchSubnet` of OVN-Kubernetes by default. If this IP address range is already in use by another network, enter the following command to update it to `100.99.0.0/16`:

Reviewer 3: ```suggestion
** `100.88.0.0/16`. This IP address range is used for the OVN-Kubernetes internal join subnet by default. If this IP address range is already in use by another network, enter the following command to update it to `100.99.0.0/16`:
```

> ** `100.88.0.0/16`. This IP address range is used for `internalTransSwitchSubnet` of OVN-Kubernetes by default. If this IP address range is already in use, enter the following command to update it to `100.99.0.0/16`:

Reviewer 4: Suggest update it to be` If this IP address range is already in use in OpenShift SDN`

Reviewer: OVN-K shall not use this subnet if it's already used by any of the customer's networks, not just by OpenShiftSDN.

+
[source,terminal]
----
$ oc patch network.operator.openshift.io cluster --type='merge' -p='{"spec":{"defaultNetwork":{"ovnKubernetesConfig":{"ipv4":{"internalTransitSwitchSubnet": "100.99.0.0/16"}}}}}'
----

> * In most cases, the live migration is independent to the secondary interfaces of pods created by the Multus CNI plugin. However, if these secondary interfaces were set up on the default network interface controller (NIC) of the host, for example, using macvlan or bridge interfaces with the default NIC as the control node, OVN-Kubernetes might encounter malfunctions. Users should remove such configurations before proceeding with the live migration.

Reviewer 4: @pliurh  About the statement of `using macvlan or bridge interfaces with the default NIC as the control nod`, how about the ipvlan?

Reviewer: Yeah, ipvlan and SRIOV shall also be mentioned here.

Author: Updated to 

```suggestion
* In most cases, the live migration is independent of the secondary interfaces of pods created by the Multus CNI plugin. However, if these secondary interfaces were set up on the default network interface controller (NIC) of the host, for example, using MACVLAN, IPVLAN, SR-IOV, or bridge interfaces with the default NIC as the control node, OVN-Kubernetes might encounter malfunctions. Users should remove such configurations before proceeding with the live migration.
```

* When there are multiple NICs inside of the host, and the default route is not on the interface that has the Kubernetes NodeIP, you must use the offline migration instead.

> * All DaemonSets in the `openshift-sdn` namespace, which are not managed by the Cluster Network Operator (CNO), must be removed. These unmanaged DaemonSets can cause the migration status to remain incomplete if not properly handled.

Reviewer 4: About the statement of `All DaemonSets in the openshift-sdn namespace, which are not managed by the Cluster Network Operator (CNO), must be removed`, should we say must be removed after the migration?

Reviewer: Any unmanaged daemonset shall be removed before the migration.

Author: Revised to: 

```suggestion
* All `DaemonSet` objects in the `openshift-sdn` namespace, which are not managed by the Cluster Network Operator (CNO), must be removed before initiating the live migration. These unmanaged DaemonSets can cause the migration status to remain incomplete if not properly handled.
* ```

> * All `DaemonSet` objects in the `openshift-sdn` namespace, which are not managed by the Cluster Network Operator (CNO), must be removed before initiating the live migration. These unmanaged DaemonSets can cause the migration status to remain incomplete if not properly handled.

Reviewer 3: ```suggestion
* All `DaemonSet` objects in the `openshift-sdn` namespace, which are not managed by the Cluster Network Operator (CNO), must be removed before initiating the live migration. These unmanaged daemon sets can cause the migration status to remain incomplete if not properly handled.
```
